terratorch/models/smp_model_factory.py

```diff
@@ -1,83 +1,387 @@
+            num_frames (int, optional): Specifies the number of timesteps the model should handle. Useful
+                        for temporal models.
+            regression_relu (bool): Whether to apply ReLU activation in the case of regression tasks.
+            **kwargs: Additional arguments that might be passed to further customize the backbone, decoder,
+                        or any auxiliary heads. These should be prefixed appropriately
+
+        Raises:
+            ValueError: If the specified decoder is not supported by SMP.
+            Exception: If the specified task is not "segmentation" or "regression".
 
         Returns:
-            Model: SMP model wrapped in SMPModelWrapper.
+            nn.Module: A model instance wrapped in SMPModelWrapper configured according to the specified
+                    parameters and tasks.
         """
-        if task not in ["segmentation", "regression"]:
-            msg = f"SMP models can only perform pixel wise tasks, but got task {task}"
-            raise Exception(msg)
-        # backbone_kwargs = _extract_prefix_keys(kwargs, "backbone_")
+
+        bands = [HLSBands.try_convert_to_hls_bands_enum(b) for b in bands]
+        if in_channels is None:
+            in_channels = len(bands)
+
+        # Gets decoder module.
+        model_module = getattr(smp, model, None)
+        if model_module is None:
+            msg = f"Decoder {model} is not supported in SMP."
+            raise ValueError(msg)
+
+        backbone_kwargs = _extract_prefix_keys(kwargs, "backbone_")  # Encoder params should be prefixed backbone_
+        smp_kwargs = _extract_prefix_keys(backbone_kwargs, "smp_")  # Smp model params should be prefixed smp_
+        aux_params = _extract_prefix_keys(backbone_kwargs, "aux_")  # Auxiliary head params should be prefixed aux_
+        aux_params = None if aux_params == {} else aux_params
+
         if isinstance(pretrained, bool):
             if pretrained:
                 pretrained = "imagenet"
             else:
                 pretrained = None
-        if decoder == "unet":
-            model = smp.Unet(
-                encoder_name=backbone, encoder_weights=pretrained, in_channels=in_channels, classes=num_classes
-            )
+
+        # If encoder not currently supported by SMP (custom encoder).
+        if backbone not in smp_encoders:
+            # These params must be included in the config file with appropriate prefix.
+            required_params = {
+                "encoder_depth": smp_kwargs,
+                "out_channels": backbone_kwargs,
+                "output_stride": backbone_kwargs,
+            }
+
+            for param, config_dict in required_params.items():
+                if param not in config_dict:
+                    msg = f"Config must include the '{param}' parameter"
+                    raise ValueError(msg)
+
+            # Using new encoder.
+            backbone_class = self._make_smp_encoder(backbone)
```

Comment:
is this working? `_make_smp_encoder` seems to be a module level method rather than a method of this class